pkg/kvstore/allocator/allocator.go

```diff
@@ -380,21 +383,32 @@ var (
 	idRandomizerMutex lock.Mutex
 )
 
-// Naive ID allocation mechanism.
-func (a *Allocator) selectAvailableID() (ID, string) {
+func (a *Allocator) newRandomIDs() {
 	idRandomizerMutex.Lock()
-	defer idRandomizerMutex.Unlock()
+	a.randomIDs = idRandomizer.Perm(int(a.max - a.min + 1))
+	idRandomizerMutex.Unlock()
+}
 
+// Naive ID allocation mechanism.
+func (a *Allocator) selectAvailableID() (ID, string) {
 	a.mutex.RLock()
 	defer a.mutex.RUnlock()
 
-	tried := 0
+	for attempt := 0; attempt < 2; attempt++ {
+		for i, r := range a.randomIDs {
+			id := ID(r) + a.min
+			if _, ok := a.cache[id]; !ok && a.localKeys.lookupID(id) == "" {
+				// remove the previously tried IDs that are already in
+				// use from the list of IDs to attempt allocation
+				if len(a.randomIDs) > 0 {
+					a.randomIDs = a.randomIDs[i+1:]
+				}
+				return id, id.String()
+			}
+		}
 
-	for _, r := range idRandomizer.Perm(int(a.max - a.min + 1)) {
-		id := ID(r) + a.min
-		tried++
-		if _, ok := a.cache[id]; !ok && a.localKeys.lookupID(id) == "" {
-			return id, id.String()
+		if attempt == 0 {
```

Comment:
We're doing 2 attempts here. If the 2nd attempt fails, this method returns leaving `a.randomIDs` empty. So in the next call to `selectAvailableID()`, the first attempt will iterate over an empty slice and do nothing, so only the 2nd attempt will do anything.
I would replace this block with a check before the 2nd for loop above:
```
if len(a.randomIDs) == 0 {
    a.newRandomIDs()
}
for i, r := range a.randomIDs {
...
}
```

Reply:
That's intentional. If we start out with an empty slice, the 2nd attempt covers a random sequence over the entire available ID space. I can leave a comment if that is not clear.

Reply:
Added comment